File: gremlin-js/gremlin-javascript/lib/structure/io/binary/internals/GraphSerializer.js
/*
 *  Licensed to the Apache Software Foundation (ASF) under one
 *  or more contributor license agreements.  See the NOTICE file
 *  distributed with this work for additional information
 *  regarding copyright ownership.  The ASF licenses this file
 *  to you under the Apache License, Version 2.0 (the
 *  "License"); you may not use this file except in compliance
 *  with the License.  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing,
 *  software distributed under the License is distributed on an
 *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 *  KIND, either express or implied.  See the License for the
 *  specific language governing permissions and limitations
 *  under the License.
 */

import { Buffer } from 'buffer';
import { Graph, Vertex, Edge, VertexProperty } from '../../../graph.js';

export default class GraphSerializer {
  constructor(ioc) {
    this.ioc = ioc;
    this.ioc.serializers[ioc.DataType.GRAPH] = this;
  }

  canBeUsedFor(value) {
    return value instanceof Graph;
  }

  serialize(item, fullyQualifiedFormat = true) {
    if (item === undefined || item === null) {
      if (fullyQualifiedFormat) {
        return Buffer.from([this.ioc.DataType.GRAPH, 0x01]);
      }
      // value-only null fallback: zero vertices + zero edges
      const zeroInt = [0x00, 0x00, 0x00, 0x00];
      return Buffer.from([...zeroInt, ...zeroInt]);
    }

    const bufs = [];
    if (fullyQualifiedFormat) {
      bufs.push(Buffer.from([this.ioc.DataType.GRAPH, 0x00]));
    }

    const vertices = item.vertices ? Array.from(item.vertices.values()) : [];
    const edges = item.edges ? Array.from(item.edges.values()) : [];

    // {vertex_count}
    bufs.push(this.ioc.intSerializer.serialize(vertices.length, false));

    // vertices
    for (const v of vertices) {
      // {id}
      bufs.push(this.ioc.anySerializer.serialize(v.id));

      // {label} as 1-element list (value-only)
      bufs.push(this.ioc.listSerializer.serialize([v.label], false));

      const vps = Array.isArray(v.properties) ? v.properties : [];

      // {vp_count}
      bufs.push(this.ioc.intSerializer.serialize(vps.length, false));

      for (const vp of vps) {
        // {vp_id}
        bufs.push(this.ioc.anySerializer.serialize(vp.id));

        // {vp_label} as 1-element list (value-only)
        bufs.push(this.ioc.listSerializer.serialize([vp.label], false));

        // {vp_value}
        bufs.push(this.ioc.anySerializer.serialize(vp.value));

        // {parent} (always null)
        bufs.push(this.ioc.unspecifiedNullSerializer.serialize(null));

        // {meta_props} as value-only List<Property>
        const metaProps = Array.isArray(vp.properties) ? vp.properties : [];
        bufs.push(this.ioc.listSerializer.serialize(metaProps, false));
      }
    }

    // {edge_count}
    bufs.push(this.ioc.intSerializer.serialize(edges.length, false));

    // edges
    for (const e of edges) {
      // {id}
      bufs.push(this.ioc.anySerializer.serialize(e.id));

      // {label} as 1-element list (value-only)
      bufs.push(this.ioc.listSerializer.serialize([e.label], false));

      // {inV_id}
      bufs.push(this.ioc.anySerializer.serialize(e.inV && e.inV.id));

      // {inV_label} (always null placeholder, fully-qualified)
      bufs.push(this.ioc.unspecifiedNullSerializer.serialize(null));

      // {outV_id}
      bufs.push(this.ioc.anySerializer.serialize(e.outV && e.outV.id));

      // {outV_label} (always null placeholder, fully-qualified)
      bufs.push(this.ioc.unspecifiedNullSerializer.serialize(null));

      // {parent} (always null)
      bufs.push(this.ioc.unspecifiedNullSerializer.serialize(null));

      // {edge_props} as value-only List<Property>
      const props = Array.isArray(e.properties) ? e.properties : [];
      bufs.push(this.ioc.listSerializer.serialize(props, false));
    }

    return Buffer.concat(bufs);
  }

  /**
   * Async deserialization of graph value bytes from a StreamReader.
   * @param {StreamReader} reader
   * @param {number} valueFlag
   * @param {number} typeCode
   * @returns {Promise<Graph>}
   */
  async deserializeValue(reader, valueFlag, typeCode) {
    const graph = new Graph();

    // {vertex_count} bare int
    const vertexCount = await this.ioc.intSerializer.deserializeBare(reader);
    for (let i = 0; i < vertexCount; i++) {
      // {id} fully qualified
      const vId = await this.ioc.anySerializer.deserialize(reader);

      // {label} value-only list, first element
      const vLabelList = await this.ioc.listSerializer.deserializeValue(reader, 0x00, this.ioc.DataType.LIST);
      const vLabel = Array.isArray(vLabelList) && vLabelList.length > 0 ? vLabelList[0] : vLabelList;

      const vertex = new Vertex(vId, vLabel, []);
      graph.vertices.set(vId, vertex);

      // {vp_count} bare int
      const vpCount = await this.ioc.intSerializer.deserializeBare(reader);
      for (let j = 0; j < vpCount; j++) {
        // {vp_id} fully qualified
        const vpId = await this.ioc.anySerializer.deserialize(reader);

        // {vp_label} value-only list, first element
        const vpLabelList = await this.ioc.listSerializer.deserializeValue(reader, 0x00, this.ioc.DataType.LIST);
        const vpLabel = Array.isArray(vpLabelList) && vpLabelList.length > 0 ? vpLabelList[0] : vpLabelList;

        // {vp_value} fully qualified
        const vpValue = await this.ioc.anySerializer.deserialize(reader);

        // {parent} fully qualified (always null)
        await this.ioc.anySerializer.deserialize(reader);

        // {meta_props} value-only list
        const metaProps = await this.ioc.listSerializer.deserializeValue(reader, 0x00, this.ioc.DataType.LIST);

        const vp = new VertexProperty(vpId, vpLabel, vpValue, metaProps || []);
        vertex.properties.push(vp);
      }
    }

    // {edge_count} bare int
    const edgeCount = await this.ioc.intSerializer.deserializeBare(reader);
    for (let i = 0; i < edgeCount; i++) {
      // {id} fully qualified
      const eId = await this.ioc.anySerializer.deserialize(reader);

      // {label} value-only list, first element
      const eLabelList = await this.ioc.listSerializer.deserializeValue(reader, 0x00, this.ioc.DataType.LIST);
      const eLabel = Array.isArray(eLabelList) && eLabelList.length > 0 ? eLabelList[0] : eLabelList;

      // {inV_id} fully qualified
      const inVId = await this.ioc.anySerializer.deserialize(reader);

      // {inV_label} fully qualified (always null placeholder) — discard
      await this.ioc.anySerializer.deserialize(reader);

      // {outV_id} fully qualified
      const outVId = await this.ioc.anySerializer.deserialize(reader);

      // {outV_label} fully qualified (always null placeholder) — discard
      await this.ioc.anySerializer.deserialize(reader);

      // {parent} fully qualified (always null) — discard
      await this.ioc.anySerializer.deserialize(reader);

      // {edge_props} value-only list
      const edgeProps = await this.ioc.listSerializer.deserializeValue(reader, 0x00, this.ioc.DataType.LIST);

      // Reuse vertex instances already in graph.vertices, otherwise build stand-ins
      const inV = graph.vertices.get(inVId) || new Vertex(inVId, '', []);
      const outV = graph.vertices.get(outVId) || new Vertex(outVId, '', []);

      const edge = new Edge(eId, outV, eLabel, inV, edgeProps || []);
      graph.edges.set(eId, edge);
      graph._indexEdge(edge);
    }

    return graph;
  }

  /**
   * Async fully-qualified deserialization from a StreamReader.
   * @param {StreamReader} reader
   * @returns {Promise<Graph|null>}
   */
  async deserialize(reader) {
    const type_code = await reader.readUInt8();
    if (type_code !== this.ioc.DataType.GRAPH) {
      throw new Error(`GraphSerializer: unexpected {type_code}=0x${type_code.toString(16)}`);
    }
    const value_flag = await reader.readUInt8();
    if (value_flag === 0x01) {
      return null;
    }
    if (value_flag !== 0x00) {
      throw new Error(`GraphSerializer: unexpected {value_flag}=0x${value_flag.toString(16)}`);
    }
    return this.deserializeValue(reader, value_flag, type_code);
  }
}

File: gremlin-dotnet/src/Gremlin.Net/Structure/IO/GraphBinary4/Types/GraphSerializer.cs
#region License

/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

#endregion

using System;
using System.Collections;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;

namespace Gremlin.Net.Structure.IO.GraphBinary4.Types
{
    /// <summary>
    /// A <see cref="Graph"/> serializer for GraphBinary. The wire format is a count-prefixed
    /// list of vertices (each with their vertex properties and meta-properties), followed by a
    /// count-prefixed list of edges (each with their properties). Vertex/edge labels are written
    /// as a single-element list, parent placeholders are written as <c>null</c>.
    /// </summary>
    public class GraphSerializer : SimpleTypeSerializer<Graph>
    {
        /// <summary>
        ///     Initializes a new instance of the <see cref="GraphSerializer" /> class.
        /// </summary>
        public GraphSerializer() : base(DataType.Graph)
        {
        }

        /// <inheritdoc />
        protected override async Task WriteValueAsync(Graph value, Stream stream, GraphBinaryWriter writer,
            CancellationToken cancellationToken = default)
        {
            await writer.WriteNonNullableValueAsync(value.Vertices.Count, stream, cancellationToken)
                .ConfigureAwait(false);
            foreach (var vertex in value.Vertices.Values)
            {
                await WriteVertexAsync(vertex, stream, writer, cancellationToken).ConfigureAwait(false);
            }

            await writer.WriteNonNullableValueAsync(value.Edges.Count, stream, cancellationToken)
                .ConfigureAwait(false);
            foreach (var edge in value.Edges.Values)
            {
                await WriteEdgeAsync(edge, stream, writer, cancellationToken).ConfigureAwait(false);
            }
        }

        private static async Task WriteVertexAsync(Vertex vertex, Stream stream, GraphBinaryWriter writer,
            CancellationToken cancellationToken)
        {
            await writer.WriteAsync(vertex.Id, stream, cancellationToken).ConfigureAwait(false);
            await writer.WriteNonNullableValueAsync(new List<string>(vertex.Labels), stream, cancellationToken)
                .ConfigureAwait(false);

            var vertexProperties = AsList<VertexProperty>(vertex.Properties);
            await writer.WriteNonNullableValueAsync(vertexProperties.Count, stream, cancellationToken)
                .ConfigureAwait(false);
            foreach (var vp in vertexProperties)
            {
                await writer.WriteAsync(vp.Id, stream, cancellationToken).ConfigureAwait(false);
                await writer.WriteNonNullableValueAsync(new List<string> { vp.Label }, stream, cancellationToken)
                    .ConfigureAwait(false);
                await writer.WriteAsync((object?)vp.Value, stream, cancellationToken).ConfigureAwait(false);

                // placeholder for the parent vertex
                await writer.WriteAsync(null, stream, cancellationToken).ConfigureAwait(false);

                var metaProperties = AsList<Property>(vp.Properties);
                await writer.WriteNonNullableValueAsync(metaProperties, stream, cancellationToken)
                    .ConfigureAwait(false);
            }
        }

        private static async Task WriteEdgeAsync(Edge edge, Stream stream, GraphBinaryWriter writer,
            CancellationToken cancellationToken)
        {
            await writer.WriteAsync(edge.Id, stream, cancellationToken).ConfigureAwait(false);
            await writer.WriteNonNullableValueAsync(new List<string>(edge.Labels), stream, cancellationToken)
                .ConfigureAwait(false);

            await writer.WriteAsync(edge.InV.Id, stream, cancellationToken).ConfigureAwait(false);
            // placeholder for the in-vertex label (always null in this context)
            await writer.WriteAsync(null, stream, cancellationToken).ConfigureAwait(false);

            await writer.WriteAsync(edge.OutV.Id, stream, cancellationToken).ConfigureAwait(false);
            // placeholder for the out-vertex label (always null in this context)
            await writer.WriteAsync(null, stream, cancellationToken).ConfigureAwait(false);

            // placeholder for the parent (never present)
            await writer.WriteAsync(null, stream, cancellationToken).ConfigureAwait(false);

            var edgeProperties = AsList<Property>(edge.Properties);
            await writer.WriteNonNullableValueAsync(edgeProperties, stream, cancellationToken)
                .ConfigureAwait(false);
        }

        /// <inheritdoc />
        protected override async Task<Graph> ReadValueAsync(Stream stream, GraphBinaryReader reader,
            CancellationToken cancellationToken = default)
        {
            var graph = new Graph();

            var vertexCount =
                (int)await reader.ReadNonNullableValueAsync<int>(stream, cancellationToken).ConfigureAwait(false);
            for (var i = 0; i < vertexCount; i++)
            {
                var vertex = await ReadVertexAsync(stream, reader, cancellationToken).ConfigureAwait(false);
                if (vertex.Id != null)
                {
                    graph.Vertices[vertex.Id] = vertex;
                }
            }

            var edgeCount =
                (int)await reader.ReadNonNullableValueAsync<int>(stream, cancellationToken).ConfigureAwait(false);
            for (var i = 0; i < edgeCount; i++)
            {
                var edge = await ReadEdgeAsync(graph, stream, reader, cancellationToken).ConfigureAwait(false);
                if (edge.Id != null)
                {
                    graph.Edges[edge.Id] = edge;
                }
            }

            return graph;
        }

        private static async Task<Vertex> ReadVertexAsync(Stream stream, GraphBinaryReader reader,
            CancellationToken cancellationToken)
        {
            var vId = await reader.ReadAsync(stream, cancellationToken).ConfigureAwait(false);
            var vLabelList = (List<string?>)await reader
                .ReadNonNullableValueAsync<List<string?>>(stream, cancellationToken).ConfigureAwait(false);
            var vLabel = vLabelList.Count > 0 ? vLabelList[0] ?? "" : "";

            var vpCount = (int)await reader.ReadNonNullableValueAsync<int>(stream, cancellationToken)
                .ConfigureAwait(false);
            var vertexProperties = new List<VertexProperty>(vpCount);
            for (var j = 0; j < vpCount; j++)
            {
                var vpId = await reader.ReadAsync(stream, cancellationToken).ConfigureAwait(false);
                var vpLabelList = (List<string?>)await reader
                    .ReadNonNullableValueAsync<List<string?>>(stream, cancellationToken).ConfigureAwait(false);
                var vpLabel = vpLabelList.Count > 0 ? vpLabelList[0] ?? "" : "";
                var vpValue = await reader.ReadAsync(stream, cancellationToken).ConfigureAwait(false);

                // discard the parent vertex placeholder
                await reader.ReadAsync(stream, cancellationToken).ConfigureAwait(false);

                var metaProps = await reader.ReadNonNullableValueAsync<List<object?>>(stream, cancellationToken)
                    .ConfigureAwait(false);
                var metaPropsArray = (metaProps as List<object>)?.ToArray() ?? Array.Empty<object>();

                vertexProperties.Add(new VertexProperty(vpId, vpLabel, vpValue, null, metaPropsArray));
            }

            return new Vertex(vId, vLabel, vertexProperties.Cast<object>().ToArray());
        }

        private static async Task<Edge> ReadEdgeAsync(Graph graph, Stream stream, GraphBinaryReader reader,
            CancellationToken cancellationToken)
        {
            var eId = await reader.ReadAsync(stream, cancellationToken).ConfigureAwait(false);
            var eLabelList = (List<string?>)await reader
                .ReadNonNullableValueAsync<List<string?>>(stream, cancellationToken).ConfigureAwait(false);
            var eLabel = eLabelList.Count > 0 ? eLabelList[0] ?? "" : "";

            var inVId = await reader.ReadAsync(stream, cancellationToken).ConfigureAwait(false);
            // discard the in-vertex label placeholder (always null on the wire)
            await reader.ReadAsync(stream, cancellationToken).ConfigureAwait(false);
            var outVId = await reader.ReadAsync(stream, cancellationToken).ConfigureAwait(false);
            // discard the out-vertex label placeholder (always null on the wire)
            await reader.ReadAsync(stream, cancellationToken).ConfigureAwait(false);
            // discard the parent placeholder
            await reader.ReadAsync(stream, cancellationToken).ConfigureAwait(false);

            var edgeProps = await reader.ReadNonNullableValueAsync<List<object?>>(stream, cancellationToken)
                .ConfigureAwait(false);
            var edgePropsArray = (edgeProps as List<object>)?.ToArray() ?? Array.Empty<object>();

            var inVertex = ResolveVertex(graph, inVId);
            var outVertex = ResolveVertex(graph, outVId);

            return new Edge(eId, outVertex, eLabel, inVertex, edgePropsArray);
        }

        private static Vertex ResolveVertex(Graph graph, object? vertexId)
        {
            if (vertexId != null && graph.Vertices.TryGetValue(vertexId, out var existing))
            {
                return existing;
            }
            return new Vertex(vertexId, "");
        }

        private static List<T> AsList<T>(IEnumerable? source)
        {
            if (source == null)
            {
                return new List<T>();
            }
            return source.Cast<T>().ToList();
        }
    }
}
